core/src/main/scala/kafka/network/SocketServer.scala

```diff
@@ -1248,6 +1257,8 @@ private[kafka] class Processor(
       if (address != null)
         connectionQuotas.dec(listenerName, address)
       selector.close(connectionId)
+      // Call listeners to notify for closed connection.
+      connectionDisconnectListeners.foreach(listener => CoreUtils.swallow(() -> listener.onDisconnect(connectionId), this, Level.ERROR))
```

Comment:
The connection ID is not a UUID, so the connection ID generated by processor_a could be the same as the one generated by processor_b. However, in the follow-up PR #17474, all processors' connection IDs are stored in a single clientConnectionIdMap. How do we handle connection ID conflicts in that case? This question may not be directly related to this PR, but it would be great to discuss it early.

Reply:
Yeah connectionId is not UUID but they are unique. The format they have is "<kafka_host_IP>-<kafka_port>-<remote_host_ip>-<remote_port>-<montonically increasing numer>". If this won't be unique then we would have bigger issues as Selector.java also keeps a map of connectionId to KafkaChannel which mutes or disconnect channels based on this connectionId.

Reply:
> The format they have is "<kafka_host_IP>-<kafka_port>-<remote_host_ip>-<remote_port>-<montonically_increasing_numer>". If this won't be unique then we would have bigger issues as Selector.java also keeps a map of connectionId to KafkaChannel which mutes or disconnect channels based on this connectionId.

Yes, but the ID is only unique within the scope of a single process. Since `monotonically_increasing_number` is a local variable of each processor, it's possible for two processors to generate the same ID. Additionally, processors don't share the same selector, so the ID doesn't need to be globally unique.

My concern was the duplicate id will corrupt the cache of (id -> UUID)

Reply:
That's a good point. I digged more into the SocketServer code and seems to me that Processors are attached to Acceptors. There could be control plane or data plane acceptors which are instantiated as per server `listeners` configuration.
So that means, there could be a possibility that if: 2 different clients + on different acceptors + with same remote host and port + generates same counter then can have same connection id. Hence do you think that we should also have listener name and security protocol passed to disconnect listeners. This information is also available in RequestContext. Or listenerName alone will be sufficient as like `connectionQuotas`?

Reply:
Hmm, two different clients from the same remote host should have different remote ports, right?

Reply:
Ok. The reason that we added monotonically_increasing_number is that a client could reuse the same port after closing but before the server realizes it. So connectionId is unique within a Processor, but not necessarily unique across Processor. Perhaps we could use both connectionId and processor id. For processors under different listeners, they have different server port. So, it doesn't seem that we need to include the listener name.

Reply:
I have reverted the last commit of mine which added listener and security protocol. Also seems to me that processorId is also not required as server port will change with processor hence making connectionId unique.

Reply:
> I missed that kafka port will also change with processor hence isn't connectionId is then unique across processors as well?

Please correct me if I'm wrong.

1. `<kafka_host_IP>-<kafka_port>`: The Processor handles connections accepted by the Acceptor, so sockets handled by different processors can have the same `<kafka_host_IP>-<kafka_port>`. This means that `<kafka_host_IP>-<kafka_port>` is not unique across processors.
2. `<remote_host_ip>-<remote_port>`: As @junrao mentioned, this part is not unique either.
3. `<monotonically_increasing_number>`: This is a local incremental variable, so it is also not unique.

In short, current implementation of `connectionId` is not unique across processors.

Reply:
@chia7712 Make sense. There are 2 options:

1. Expose processor id in RequestContext so at other places, unique connection reference can be attached.
2. Add processor id in connection id i.e. `<kafka_host_IP>:<kafka_port>-<remote_host_ip>:<remote_port>-<monotonically_increasing_number>-<processor-id>`

@chia7712 @junrao I am inclined towards 2, do you see any issue in adding processor id in connection id creation?

Reply:
> <kafka_host_IP>:<kafka_port>-<remote_host_ip>:<remote_port>-<monotonically_increasing_number>-<processor-id>

Reply:
Yeah, while implementing I thought to have processor id prior from `monotonically_increasing_number` as the `monotonically_increasing_number` belongs to processor id.